main.go

```diff
@@ -231,6 +231,31 @@ func run(ctrlConfig spirev1alpha1.ControllerManagerConfig, options ctrl.Options)
 		return err
 	}
 
+	foundClusterStaticEntry := false
+	foundClusterSpiffeID := false
+	foundClusterFederatedTrustDomain := false
+
+	_, resources, err := clientset.ServerGroupsAndResources()
+	for _, r := range resources {
+		if r.GroupVersion == "spire.spiffe.io/v1alpha1" {
```

Comment:
this will disable reconciler when starting in case a CRD is not found, but what happens if it is introduced after that?
maybe it is better to keep setup reconcilers (as it is already done) and just add a warning when reconciler is called? (when list failed)

what do you think?

Reply:
I have that logic within the `spireentry.Reconciler`... however the controller still crashes eventually if you create it - I'm not sure there is a graceful way to handle that with the client. Unless you wanted to throw the initialization of the controllers in a loop.

Current proposed solution would require you to restart the controller manager if you update the CRDs

Logs from a solution that goes ahead and creates the `controllers.ClusterStaticEntryReconciler`:
```
2023-07-17T20:17:28Z	INFO	entry-reconciler	ClusterStaticEntry CRD not installed
2023-07-17T20:17:37Z	ERROR	controller-runtime.source.EventHandler	if kind is a CRD, it should be installed before calling Start	{"kind": "ClusterStaticEntry.spire.spiffe.io", "error": "no matches for kind \"ClusterStaticEntry\" in version \"spire.spiffe.io/v1alpha1\""}
sigs.k8s.io/controller-runtime/pkg/internal/source.(*Kind).Start.func1.1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/source/kind.go:63
k8s.io/apimachinery/pkg/util/wait.loopConditionUntilContext.func1
	/go/pkg/mod/k8s.io/apimachinery@v0.27.3/pkg/util/wait/loop.go:62
k8s.io/apimachinery/pkg/util/wait.loopConditionUntilContext
	/go/pkg/mod/k8s.io/apimachinery@v0.27.3/pkg/util/wait/loop.go:63
k8s.io/apimachinery/pkg/util/wait.PollUntilContextCancel
	/go/pkg/mod/k8s.io/apimachinery@v0.27.3/pkg/util/wait/poll.go:33
sigs.k8s.io/controller-runtime/pkg/internal/source.(*Kind).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/source/kind.go:56
2023-07-17T20:17:38Z	INFO	entry-reconciler	ClusterStaticEntry CRD not installed
2023-07-17T20:17:47Z	ERROR	Could not wait for Cache to sync	{"controller": "clusterstaticentry", "controllerGroup": "spire.spiffe.io", "controllerKind": "ClusterStaticEntry", "error": "failed to wait for clusterstaticentry caches to sync: timed out waiting for cache to be synced for Kind *v1alpha1.ClusterStaticEntry"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:202
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:207
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:233
sigs.k8s.io/controller-runtime/pkg/manager.(*runnableGroup).reconcile.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/manager/runnable_group.go:219
2023-07-17T20:17:47Z	INFO	Stopping and waiting for non leader election runnables
2023-07-17T20:17:47Z	INFO	shutting down server	{"path": "/metrics", "kind": "metrics", "addr": "[::]:8082"}
2023-07-17T20:17:47Z	INFO	Stopping and waiting for leader election runnables
2023-07-17T20:17:47Z	INFO	entry-reconciler	Reconciliation canceled
2023-07-17T20:17:47Z	ERROR	error received after stop sequence was engaged	{"error": "context canceled"}
sigs.k8s.io/controller-runtime/pkg/manager.(*controllerManager).engageStopProcedure.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/manager/internal.go:555
2023-07-17T20:17:47Z	INFO	federation relationship-reconciler	Reconciliation canceled
2023-07-17T20:17:47Z	ERROR	error received after stop sequence was engaged	{"error": "context canceled"}
sigs.k8s.io/controller-runtime/pkg/manager.(*controllerManager).engageStopProcedure.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/manager/internal.go:555
2023-07-17T20:17:47Z	INFO	Shutdown signal received, waiting for all workers to finish	{"controller": "pod", "controllerGroup": "", "controllerKind": "Pod"}
2023-07-17T20:17:47Z	INFO	Shutdown signal received, waiting for all workers to finish	{"controller": "clusterspiffeid", "controllerGroup": "spire.spiffe.io", "controllerKind": "ClusterSPIFFEID"}
2023-07-17T20:17:47Z	INFO	Shutdown signal received, waiting for all workers to finish	{"controller": "clusterfederatedtrustdomain", "controllerGroup": "spire.spiffe.io", "controllerKind": "ClusterFederatedTrustDomain"}
2023-07-17T20:17:47Z	ERROR	error received after stop sequence was engaged	{"error": "context canceled"}
sigs.k8s.io/controller-runtime/pkg/manager.(*controllerManager).engageStopProcedure.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/manager/internal.go:555
2023-07-17T20:17:47Z	INFO	All workers finished	{"controller": "pod", "controllerGroup": "", "controllerKind": "Pod"}
2023-07-17T20:17:47Z	INFO	All workers finished	{"controller": "clusterfederatedtrustdomain", "controllerGroup": "spire.spiffe.io", "controllerKind": "ClusterFederatedTrustDomain"}
2023-07-17T20:17:47Z	INFO	All workers finished	{"controller": "clusterspiffeid", "controllerGroup": "spire.spiffe.io", "controllerKind": "ClusterSPIFFEID"}
2023-07-17T20:17:47Z	INFO	Stopping and waiting for caches
2023-07-17T20:17:47Z	INFO	Stopping and waiting for webhooks
2023-07-17T20:17:47Z	INFO	controller-runtime.webhook	Shutting down webhook server with timeout of 1 minute
2023-07-17T20:17:47Z	INFO	Wait completed, proceeding to shutdown the manager
2023-07-17T20:17:47Z	ERROR	setup	problem running manager	{"error": "failed to wait for clusterstaticentry caches to sync: timed out waiting for cache to be synced for Kind *v1alpha1.ClusterStaticEntry"}
main.run
	/workspace/main.go:377
main.main
	/workspace/main.go:81
runtime.main
	/usr/local/go/src/runtime/proc.go:250
```